Integrate dpdk runmode v8

lukashino opened the pull request:
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/issues/4507) ticket:

Describe changes:
- Integrate DPDK capture interface as an alternative to AF_PACKET or others.
- Focus implementation on workers run mode.
- Support both IDS and IPS modes.
- Follow AF_PACKET capture interface architecture and apply DPDK best practices for development.
- Update documentation to reflect new changes.
- Edit Github workflow to include build for the DPDK capture interface.

Work relies on pkg-config and requires at least DPDK 19.11. Work is focused on adding support for physical NICs. At the same time, it should be run natively (and not in docker or in virtual machines) as a primary process. To support multiple workers, the NICs need to support RSS offload.

Changelog from the previous [PR](https://github.com/OISF/suricata/pull/6527):
- added RSS support for the driver i40e for DPDK 19.11+
- updated documentation to reflect comments from the previous PR
- updated suricata.yaml.in - the default config
- relaxed conditions when loading the configuration from the YAML file 
- refactoring

victorjulien commented:
```
[862309] 14/11/2021 -- 08:53:35 - (runmode-dpdk.c:828) <Warning> (DeviceConfigureQueues) -- [ERRCODE: SC_ERR_PCRE_COMPILE(5)] - Thread statistics can be incomplete. Set number of threads to 16 or increase RTE_ETHDEV_QUEUE_STAT_CNTRS when compiling DPDK
```
This is with `threads: auto`. What is the consequence of this? Should `auto` use a max value of `RTE_ETHDEV_QUEUE_STAT_CNTRS`?

victorjulien commented:
So on my Intel x520 setup, I get a warning by default:
```
[862666] 14/11/2021 -- 09:42:35 - (runmode-dpdk.c:776) <Warning> (DeviceInitPortConf) -- [ERRCODE: SC_WARN_DPDK_CONF(344)] - Interface 0000:05:00.1 modified RSS hash function based on hardware support, requested:0xa38c configured:0x8104
```
I find this message to be unclear, and also I would like the default config to not generate a warning at all. If it is because the ixgbe based NICs don't support certain features, can we just set things up so that things are supported?

See also https://github.com/OISF/suricata/pull/6527#discussion_r748414449

victorjulien commented on src/source-dpdk.c:
```diff
@@ -0,0 +1,617 @@
+     * the port level. Therefore setting it to the first worker to have at least continuous update
+     * on the dropped packets. */
+    if (ptv->queue_id == 0) {
+        StatsSetUI64(ptv->tv, ptv->capture_dpdk_rx_errs,
```
I think we should get the drop counter back here. `rx_errors` seems to include it, but we need to follow the other methods on this.

I am thinking about capture method specific stats. Perhaps we can have `capture.dpdk.rx_nombuf`, and such.

victorjulien replied:
This is probably all drops, but its hard to know for sure
```
Date: 11/14/2021 -- 13:52:45 (uptime: 0d, 04h 11m 41s)
------------------------------------------------------------------------------------
Counter                                       | TM Name                   | Value
------------------------------------------------------------------------------------
capture.packets                               | Total                     | 11181144783
capture.rx_errors                             | Total                     | 48889667119
```

victorjulien replied:
It looks like `capture.packets` does not include the `dropped` packets, which is how it is in afpacket. So I guess `capture.packets` should include `imissed` and possible other errors?

lukashino replied:
So, do you want to have one counter `capture.drop` that will include all packets that were dropped for whatever reason (whether it is RX or TX direction) and then individual counters like `capture.dpdk.rx_nombuf`, `capture.dpdk.imissed` and etc.?

lukashino replied:
> It looks like `capture.packets` does not include the `dropped` packets, which is how it is in afpacket. So I guess `capture.packets` should include `imissed` and possible other errors?

I can reimplement the logic to follow the behavior of AFP.

lukashino replied:
I have added the count of received packets to the total count of dropped packets. Here is what I currently have:
```c
if (ptv->queue_id == 0) {
        StatsSetUI64(ptv->tv, ptv->capture_dpdk_rx_errs,
                eth_stats.imissed + eth_stats.ierrors + eth_stats.rx_nombuf + ptv->pkts);
        StatsSetUI64(ptv->tv, ptv->capture_dpdk_tx_errs, eth_stats.oerrors);
        SC_ATOMIC_SET(ptv->livedev->drop,
                eth_stats.imissed + eth_stats.ierrors + eth_stats.rx_nombuf + eth_stats.oerrors + ptv->pkts);
    }
```

What I suggest is the following:
- remove the counter ptv->capture_dpdk_rx_errs and have independent counters for each kind of error
- the counter `ptv->livedev->drop` should only include packets that were received or failed to receive (it should not include `eth_stats.oerrors`) -> I haven't seen in AFP capture interface that it counts packets that were failed to a dropped (or any other) counter

Do you agree with the proposed solution?

victorjulien commented on suricata.yaml.in:
```diff
@@ -649,6 +649,63 @@ af-packet:
+      mtu: 1500 # Set MTU of the device in bytes
+
+      # To approximately calculate required amount of space (in bytes) for interface's mempool: mempool-size * mtu
+      # Make sure you have enough allocated hugepages.
```
Can we calculate how many hugepages we need? Additionally, could we then check that we have enough at start up?

lukashino replied:
I am not aware of a library that could retrieve the amount and size of individual hugepages. Of course, it is possible to do a call to OS but other than that I can't think of anything better.

victorjulien commented:
> So on my Intel x520 setup, I get a warning by default:
> 
> ```
> [862666] 14/11/2021 -- 09:42:35 - (runmode-dpdk.c:776) <Warning> (DeviceInitPortConf) -- [ERRCODE: SC_WARN_DPDK_CONF(344)] - Interface 0000:05:00.1 modified RSS hash function based on hardware support, requested:0xa38c configured:0x8104
> ```
> 
> I find this message to be unclear, and also I would like the default config to not generate a warning at all. If it is because the ixgbe based NICs don't support certain features, can we just set things up so that things are supported?
> 
> See also [#6527 (comment)](https://github.com/OISF/suricata/pull/6527#discussion_r748414449)

I'm experimenting a bit to find out more about this, these are the flags requested vs actually set on my ixgbe:
```
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:773) <Notice> (DeviceInitPortConf) -- REQUESTED:
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:775) <Notice> (DeviceInitPortConf) -- ETH_RSS_IP set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:776) <Notice> (DeviceInitPortConf) -- ETH_RSS_IPV4 set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:777) <Notice> (DeviceInitPortConf) -- ETH_RSS_FRAG_IPV4 set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:778) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV4_TCP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:779) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV4_UDP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:780) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV4_SCTP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:781) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV4_OTHER set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:782) <Notice> (DeviceInitPortConf) -- ETH_RSS_IPV6 set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:783) <Notice> (DeviceInitPortConf) -- ETH_RSS_FRAG_IPV6 set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:784) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV6_TCP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:785) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV6_UDP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:786) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV6_SCTP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:787) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV6_OTHER set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:789) <Notice> (DeviceInitPortConf) -- ETH_RSS_TCP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:790) <Notice> (DeviceInitPortConf) -- ETH_RSS_UDP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:791) <Notice> (DeviceInitPortConf) -- ETH_RSS_SCTP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:793) <Notice> (DeviceInitPortConf) -- ETH_RSS_L2_PAYLOAD NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:794) <Notice> (DeviceInitPortConf) -- ETH_RSS_IPV6_EX set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:795) <Notice> (DeviceInitPortConf) -- ETH_RSS_IPV6_TCP_EX NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:796) <Notice> (DeviceInitPortConf) -- ETH_RSS_IPV6_UDP_EX NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:798) <Notice> (DeviceInitPortConf) -- ETH_RSS_PORT NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:799) <Notice> (DeviceInitPortConf) -- ETH_RSS_VXLAN NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:800) <Notice> (DeviceInitPortConf) -- ETH_RSS_NVGRE NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:801) <Notice> (DeviceInitPortConf) -- ETH_RSS_GTPU NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:803) <Notice> (DeviceInitPortConf) -- ETH_RSS_L3_SRC_ONLY NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:804) <Notice> (DeviceInitPortConf) -- ETH_RSS_L3_DST_ONLY NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:805) <Notice> (DeviceInitPortConf) -- ETH_RSS_L4_SRC_ONLY NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:806) <Notice> (DeviceInitPortConf) -- ETH_RSS_L4_DST_ONLY NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:811) <Warning> (DeviceInitPortConf) -- [ERRCODE: SC_WARN_DPDK_CONF(344)] - Interface 0000:05:00.1 modified RSS hash function based on hardware support, requested:0xa38c configured:0x8104
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:818) <Notice> (DeviceInitPortConf) -- ACTUAL:
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:819) <Notice> (DeviceInitPortConf) -- ETH_RSS_IP set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:820) <Notice> (DeviceInitPortConf) -- ETH_RSS_IPV4 set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:821) <Notice> (DeviceInitPortConf) -- ETH_RSS_FRAG_IPV4 NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:822) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV4_TCP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:823) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV4_UDP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:824) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV4_SCTP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:825) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV4_OTHER NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:826) <Notice> (DeviceInitPortConf) -- ETH_RSS_IPV6 set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:827) <Notice> (DeviceInitPortConf) -- ETH_RSS_FRAG_IPV6 NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:828) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV6_TCP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:829) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV6_UDP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:830) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV6_SCTP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:831) <Notice> (DeviceInitPortConf) -- ETH_RSS_NONFRAG_IPV6_OTHER NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:833) <Notice> (DeviceInitPortConf) -- ETH_RSS_TCP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:834) <Notice> (DeviceInitPortConf) -- ETH_RSS_UDP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:835) <Notice> (DeviceInitPortConf) -- ETH_RSS_SCTP NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:837) <Notice> (DeviceInitPortConf) -- ETH_RSS_L2_PAYLOAD NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:838) <Notice> (DeviceInitPortConf) -- ETH_RSS_IPV6_EX set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:839) <Notice> (DeviceInitPortConf) -- ETH_RSS_IPV6_TCP_EX NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:840) <Notice> (DeviceInitPortConf) -- ETH_RSS_IPV6_UDP_EX NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:842) <Notice> (DeviceInitPortConf) -- ETH_RSS_PORT NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:843) <Notice> (DeviceInitPortConf) -- ETH_RSS_VXLAN NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:844) <Notice> (DeviceInitPortConf) -- ETH_RSS_NVGRE NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:845) <Notice> (DeviceInitPortConf) -- ETH_RSS_GTPU NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:847) <Notice> (DeviceInitPortConf) -- ETH_RSS_L3_SRC_ONLY NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:848) <Notice> (DeviceInitPortConf) -- ETH_RSS_L3_DST_ONLY NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:849) <Notice> (DeviceInitPortConf) -- ETH_RSS_L4_SRC_ONLY NOT set
[863934] 14/11/2021 -- 20:22:52 - (runmode-dpdk.c:850) <Notice> (DeviceInitPortConf) -- ETH_RSS_L4_DST_ONLY NOT set
```

victorjulien commented:
Back again on my x520 (ixgbe), I see quite a few more drops than with af-packet. I wonder if it is related to the RSS settings. It seems quite unbalanced:
```
capture.packets                               | Total                     | 4299119450
capture.rx_errors                             | Total                     | 615572927
decoder.pkts                                  | Total                     | 4299152973
capture.packets                               | W#01-0000:05:00.1         | 471234445
capture.rx_errors                             | W#01-0000:05:00.1         | 615572927
decoder.pkts                                  | W#01-0000:05:00.1         | 471234807
capture.packets                               | W#02-0000:05:00.1         | 516564644
capture.rx_errors                             | W#02-0000:05:00.1         | 0
decoder.pkts                                  | W#02-0000:05:00.1         | 516565735
capture.packets                               | W#03-0000:05:00.1         | 262524276
capture.rx_errors                             | W#03-0000:05:00.1         | 0
decoder.pkts                                  | W#03-0000:05:00.1         | 262524996
capture.packets                               | W#04-0000:05:00.1         | 221024860
capture.rx_errors                             | W#04-0000:05:00.1         | 0
decoder.pkts                                  | W#04-0000:05:00.1         | 221025673
capture.packets                               | W#05-0000:05:00.1         | 243440054
capture.rx_errors                             | W#05-0000:05:00.1         | 0
decoder.pkts                                  | W#05-0000:05:00.1         | 243441336
capture.packets                               | W#06-0000:05:00.1         | 210569670
capture.rx_errors                             | W#06-0000:05:00.1         | 0
decoder.pkts                                  | W#06-0000:05:00.1         | 210571579
capture.packets                               | W#07-0000:05:00.1         | 193204934
capture.rx_errors                             | W#07-0000:05:00.1         | 0
decoder.pkts                                  | W#07-0000:05:00.1         | 193206054
capture.packets                               | W#08-0000:05:00.1         | 281573377
capture.rx_errors                             | W#08-0000:05:00.1         | 0
decoder.pkts                                  | W#08-0000:05:00.1         | 281579117
capture.packets                               | W#09-0000:05:00.1         | 270435595
capture.rx_errors                             | W#09-0000:05:00.1         | 0
decoder.pkts                                  | W#09-0000:05:00.1         | 270439858
capture.packets                               | W#10-0000:05:00.1         | 199514871
capture.rx_errors                             | W#10-0000:05:00.1         | 0
decoder.pkts                                  | W#10-0000:05:00.1         | 199515229
capture.packets                               | W#11-0000:05:00.1         | 252849473
capture.rx_errors                             | W#11-0000:05:00.1         | 0
decoder.pkts                                  | W#11-0000:05:00.1         | 252851189
capture.packets                               | W#12-0000:05:00.1         | 208507008
capture.rx_errors                             | W#12-0000:05:00.1         | 0
decoder.pkts                                  | W#12-0000:05:00.1         | 208508447
capture.packets                               | W#13-0000:05:00.1         | 188770898
capture.rx_errors                             | W#13-0000:05:00.1         | 0
decoder.pkts                                  | W#13-0000:05:00.1         | 188776766
capture.packets                               | W#14-0000:05:00.1         | 288527193
capture.rx_errors                             | W#14-0000:05:00.1         | 0
decoder.pkts                                  | W#14-0000:05:00.1         | 288531046
capture.packets                               | W#15-0000:05:00.1         | 269408985
capture.rx_errors                             | W#15-0000:05:00.1         | 0
decoder.pkts                                  | W#15-0000:05:00.1         | 269410535
capture.packets                               | W#16-0000:05:00.1         | 220969167
capture.rx_errors                             | W#16-0000:05:00.1         | 0
decoder.pkts                                  | W#16-0000:05:00.1         | 220970606
capture.packets                               | W#17-0000:05:00.1         | 0
capture.rx_errors                             | W#17-0000:05:00.1         | 0
decoder.pkts                                  | W#17-0000:05:00.1         | 0
capture.packets                               | W#18-0000:05:00.1         | 0
capture.rx_errors                             | W#18-0000:05:00.1         | 0
decoder.pkts                                  | W#18-0000:05:00.1         | 0
```
Note that there are no packets on threads 17 and 18. Perhaps we have 16 queues max? The `decoder.pkts` would have incremented even if dpdk would have a stats limit.

lukashino commented:
> Note that there are no packets on threads 17 and 18. Perhaps we have 16 queues max? The decoder.pkts would have incremented even if dpdk would have a stats limit.

Yes, that's the reason why you don't see packets on queues 17 and 18. 

Uncertain atm why it seems rather unbalanced but I would assume something is odd with the RSS settings. I'll try to have a look at it tomorrow.

lukashino commented:
Just to confirm, is your X520 (ixgbe) running on DPDK 20.xx?

victorjulien commented:
> Just to confirm, is your X520 (ixgbe) running on DPDK 20.xx?

Ya, 20.11 (Ubuntu pkg 20.11.3-0ubuntu0.21.04.2)

victorjulien commented:
> > Note that there are no packets on threads 17 and 18. Perhaps we have 16 queues max? The decoder.pkts would have incremented even if dpdk would have a stats limit.
> 
> Yes, that's the reason why you don't see packets on queues 17 and 18.

Can we get the max queues through the dpdk api and set that as the max number of threads?

victorjulien commented:
> Uncertain atm why it seems rather unbalanced but I would assume something is odd with the RSS settings. I'll try to have a look at it tomorrow.

I should say that in af-packet I was using `cluster_flow`, which does in kernel flow based load balancing. So this does not rely on RSS in my af-packet test.

lukashino commented:
> > > Note that there are no packets on threads 17 and 18. Perhaps we have 16 queues max? The decoder.pkts would have incremented even if dpdk would have a stats limit.
> > 
> > 
> > Yes, that's the reason why you don't see packets on queues 17 and 18.
> 
> Can we get the max queues through the dpdk api and set that as the max number of threads?

Do you want to limit the number of workers to the number of queues where counters are still enabled? The number of queues available with the counters are fixed from the build time constant and to my knowledge it can only be changed during the DPDK build process. Even though the stats for the given workers are not shown, they are still active and process packets. The stats of the port take in account also packets received on queues where counters are not enabled anymore. 

But when I look at your logs, I would assume the decoder.pkts would be increasing

victorjulien commented:
> > > > Note that there are no packets on threads 17 and 18. Perhaps we have 16 queues max? The decoder.pkts would have incremented even if dpdk would have a stats limit.
> > > 
> > > 
> > > Yes, that's the reason why you don't see packets on queues 17 and 18.
> > 
> > 
> > Can we get the max queues through the dpdk api and set that as the max number of threads?
> 
> Do you want to limit the number of workers to the number of queues where counters are still enabled? The number of queues available with the counters are fixed from the build time constant and to my knowledge it can only be changed during the DPDK build process. Even though the stats for the given workers are not shown, they are still active and process packets. The stats of the port take in account also packets received on queues where counters are not enabled anymore.
> 
> But when I look at your logs, I would assume at least the decoder.pkts would be increasing

I think there are two issues: one is taking the max queues into consideration (set as threads in `auto`, warn if threads > max_queues in manual threads setting), the other is that if we use `auto` it should perhaps not exceed the max number queues for which it can track stats? I see going beyond that as something that should be left to advanced users that know what they are doing.

lukashino commented:
> Back again on my x520 (ixgbe), I see quite a few more drops than with af-packet. I wonder if it is related to the RSS settings. It seems quite unbalanced:
> 
> ...
>
> Note that there are no packets on threads 17 and 18. Perhaps we have 16 queues max? The `decoder.pkts` would have incremented even if dpdk would have a stats limit.

What traffic have you been sending to Suricata and approximately at what speed? Also did you run it in IDS mode?

lukashino commented:
> > > > > Note that there are no packets on threads 17 and 18. Perhaps we have 16 queues max? The decoder.pkts would have incremented even if dpdk would have a stats limit.
> > > > 
> > > > 
> > > > Yes, that's the reason why you don't see packets on queues 17 and 18.
> > > 
> > > 
> > > Can we get the max queues through the dpdk api and set that as the max number of threads?
> > 
> > 
> > Do you want to limit the number of workers to the number of queues where counters are still enabled? The number of queues available with the counters are fixed from the build time constant and to my knowledge it can only be changed during the DPDK build process. Even though the stats for the given workers are not shown, they are still active and process packets. The stats of the port take in account also packets received on queues where counters are not enabled anymore.
> > But when I look at your logs, I would assume at least the decoder.pkts would be increasing
> 
> I think there are two issues: one is taking the max queues into consideration (set as threads in `auto`, warn if threads > max_queues in manual threads setting), the other is that if we use `auto` it should perhaps not exceed the max number queues for which it can track stats? I see going beyond that as something that should be left to advanced users that know what they are doing.

I like this approach, I can limit the `auto` setting to the number of queues able to track stats.

lukashino commented:
For the record of i40e RSS configuration evaluation. Victor have noticed that I set IP but also SCTP / TCP/ UDP/ OTHER and fragmented IP patterns (e.g. function `i40eDeviceSetRSSFlowIPv4`). Since I haven't tested receiving fragmented packets when I first submitted this PR, I've decided to give it a try. 
Tests:
1) As is in PR: 
- fragmented and nonfragmented packets are received by the same workers even when I applied seed to alter them via `tcpreplay-edit` (option `--seed`)
2) Setting only `ETH_RSS_FRAG_IPV4` and `ETH_RSS_IPV4`
- when setting `ETH_RSS_IPV4`, the PMD driver says that pctype is not supported (generally this means that the "type" of traffic is not a valid configuration for the i40e)
3) Setting only `ETH_RSS_FRAG_IPV4` and `ETH_RSS_NONFRAG_IPV4_OTHER`
- this doesn't work well, packets of the same flow are received on the different workers (my explanation is that the fragmented packets are matched with `ETH_RSS_FRAG_IPV4` but the other UDP packets are not matched with `ETH_RSS_NONFRAG_IPV4_OTHER` rte_flow rule.

Also only IPV4 version of flows were mentioned but I applied the same settings to IPv6 rules.

For this reason, I believe the configuration is correct and the RSS distributes packets as it should.

lukashino commented:
> > > > > > Note that there are no packets on threads 17 and 18. Perhaps we have 16 queues max? The decoder.pkts would have incremented even if dpdk would have a stats limit.
> > > > > 
> > > > > 
> > > > > Yes, that's the reason why you don't see packets on queues 17 and 18.
> > > > 
> > > > 
> > > > Can we get the max queues through the dpdk api and set that as the max number of threads?
> > > 
> > > 
> > > Do you want to limit the number of workers to the number of queues where counters are still enabled? The number of queues available with the counters are fixed from the build time constant and to my knowledge it can only be changed during the DPDK build process. Even though the stats for the given workers are not shown, they are still active and process packets. The stats of the port take in account also packets received on queues where counters are not enabled anymore.
> > > But when I look at your logs, I would assume at least the decoder.pkts would be increasing
> > 
> > 
> > I think there are two issues: one is taking the max queues into consideration (set as threads in `auto`, warn if threads > max_queues in manual threads setting), the other is that if we use `auto` it should perhaps not exceed the max number queues for which it can track stats? I see going beyond that as something that should be left to advanced users that know what they are doing.
> 
> I like this approach, I can limit the `auto` setting to the number of queues able to track stats.

I got confused by how the stats used to be obtained from the queues in the old PRs. Currently, queue stats are not used. That means that the warning "Thread statistics can be incomplete" is not applicable anymore (as we are getting only global port stats and the received packets are counted in each of the software workers). 
I have already removed the warning. 
That means there doesn't need to be a limitation we have talked about. There is already a check if the number of configured queues is less than the maximal number of queues. 

That also means that the culprit of the problem (no packets on workers 17, 18) must be somewhere else. I was wondering.. was your yaml configuration good? Didn't you possibly had 18queues but only 16cores set in CPU affinity settings?

I've switched over to a machine with 24 cores to verify your findings with ixgbe. I found out that:
- reason why your have so many errors on worker 1 and none on other workers is actually caused by the code because errors are fetched again only on a global (port) level and set only by the core 0
- RSS is working fine on my end. I had 20 cores enabled and below is the output from the `stats.log`. I've also tested a UDP flow with both non / fragmented packets.

```
tail -n1200 ~/suricata/var/log/suricata/stats.log | grep "capture.packets\|capture.rx_errors"
capture.packets                               | Total                                 | 1000006
capture.rx_errors                             | Total                                 | 1
capture.packets                               | W#01-0000:03:00.0         | 69733
capture.rx_errors                             | W#01-0000:03:00.0         | 1
capture.packets                               | W#02-0000:03:00.0         | 44620
capture.packets                               | W#03-0000:03:00.0         | 50974
capture.packets                               | W#04-0000:03:00.0         | 54534
capture.packets                               | W#05-0000:03:00.0         | 61601
capture.packets                               | W#06-0000:03:00.0         | 31513
capture.packets                               | W#07-0000:03:00.0         | 39905
capture.packets                               | W#08-0000:03:00.0         | 70383
capture.packets                               | W#09-0000:03:00.0         | 72525
capture.packets                               | W#10-0000:03:00.0         | 52845
capture.packets                               | W#11-0000:03:00.0         | 38961
capture.packets                               | W#12-0000:03:00.0         | 47202
capture.packets                               | W#13-0000:03:00.0         | 40466
capture.packets                               | W#14-0000:03:00.0         | 48887
capture.packets                               | W#15-0000:03:00.0         | 50675
capture.packets                               | W#16-0000:03:00.0         | 32461
capture.packets                               | W#17-0000:03:00.0         | 56261
capture.packets                               | W#18-0000:03:00.0         | 32136
capture.packets                               | W#19-0000:03:00.0         | 53866
capture.packets                               | W#20-0000:03:00.0         | 50458
```

lukashino commented on src/util-dpdk-i40e.c:
```diff
@@ -0,0 +1,348 @@
+    memset(&info, 0, sizeof(info));
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
```
The support for RSS hashing for `i40e` on DPDK 19.11 is flaged as deprecated but still gives me the best results. I've tried to support RSS configuration through rte_flow rules but I wasn't able to make it. As this is only solution for DPDK 19.xx I think the deprecation warnings can be ignored and for newer versions rte_flow can be used.

lukashino commented on src/util-dpdk-i40e.h:
```diff
@@ -0,0 +1,47 @@
+ * \author Lukas Sismis <lukas.sismis@gmail.com>
+ */
+
+#ifndef SURICATA_UTIL_DPDK_I40E_H
```
My IDE sneaked it in. Will remove.

lukashino commented on src/util-dpdk.h:
```diff
@@ -0,0 +1,48 @@
+#include <rte_flow.h>
+
+#define RSS_HKEY_LEN      40
+#define I40E_RSS_HKEY_LEN 52
```
moved

lukashino commented on src/util-dpdk.h:
```diff
@@ -0,0 +1,48 @@
+#include <rte_mbuf.h>
+#include <rte_flow.h>
+
+#define RSS_HKEY_LEN      40
```
I've moved it there.

lukashino commented on src/runmode-dpdk.c:
```diff
@@ -0,0 +1,1237 @@
+    SCReturnInt(0);
+}
+
+static int ConfigLoad(DPDKIfaceConfig *iconf, const char *iface)
```
At this function, checking if EAL parameter was specified and set to `primary` is too late. This can be done in InitEal() function in my opinion. DPDK apps start as primary processes by default. I can add the check directly there. Currently a nice thing that can be done when Suricata is running as a secondary process is to create a virtual port and send packets from the primary app there. It's something that Amaan has been doing in the previous PRs.

lukashino commented on src/runmode-dpdk.c:
```diff
@@ -0,0 +1,1237 @@
+    struct rte_eth_rxconf rxq_conf;
+    struct rte_eth_txconf txq_conf;
+
+    if (unlikely(iconf->nb_rx_queues >= RTE_ETHDEV_QUEUE_STAT_CNTRS)) {
```
removed the warning completely as queue stats are not used anymore.

lukashino commented on src/runmode-dpdk.c:
```diff
@@ -0,0 +1,1237 @@
+
+    if (strcmp(entry_str, "none") != 0 && strcmp(entry_str, "tap") != 0 &&
+            strcmp(entry_str, "ips") != 0) {
+        SCLogError(SC_ERR_INVALID_VALUE,
```
changed to warning

lukashino commented on src/util-dpdk-i40e.h:
```diff
@@ -0,0 +1,47 @@
+
+#ifdef HAVE_DPDK
+
+int i40eDeviceSetRSS(int port_id, int nb_rx_queues);
```
Done.

victorjulien commented:
> For the record of i40e RSS configuration evaluation. Victor have noticed that I set IP but also SCTP / TCP/ UDP/ OTHER and fragmented IP patterns (e.g. function `i40eDeviceSetRSSFlowIPv4`). Since I haven't tested receiving fragmented packets when I first submitted this PR, I've decided to give it a try. Tests:
> 
> 1. As is in PR:
> 
> * fragmented and nonfragmented packets are received by the same workers even when I applied seed to alter them via `tcpreplay-edit` (option `--seed`)
> 
> 1. Setting only `ETH_RSS_FRAG_IPV4` and `ETH_RSS_IPV4`
> 
> * when setting `ETH_RSS_IPV4`, the PMD driver says that pctype is not supported (generally this means that the "type" of traffic is not a valid configuration for the i40e)
> 
> 1. Setting only `ETH_RSS_FRAG_IPV4` and `ETH_RSS_NONFRAG_IPV4_OTHER`
> 
> * this doesn't work well, packets of the same flow are received on the different workers (my explanation is that the fragmented packets are matched with `ETH_RSS_FRAG_IPV4` but the other UDP packets are not matched with `ETH_RSS_NONFRAG_IPV4_OTHER` rte_flow rule.
> 
> Also only IPV4 version of flows were mentioned but I applied the same settings to IPv6 rules.
> 
> For this reason, I believe the configuration is correct and the RSS distributes packets as it should.

Can you make sure this analysis is part of the git commit messages for the i40e specific commit?

lukashino commented on src/util-dpdk-i40e.c:
```diff
@@ -0,0 +1,348 @@
+int i40eDeviceSetRSS(int port_id, int nb_rx_queues)
+{
+    (void)nb_rx_queues; // avoid unused variable warnings
+    return RTE_VER_YEAR <= 19 ? i40eDeviceSetRSSWithFilter(port_id)
```
Done! In the beginning, my point was to have a code that can be compiled on one DPDK version and used with the other version but then I got burdened by the problems you have pinpointed.

lukashino commented on src/runmode-dpdk.c:
```diff
@@ -0,0 +1,1237 @@
+            };
+            // RSS is configured in a specific way for a driver i40e and DPDK version
+            // less or equal to 19.xx
+            if (strcmp(dev_info->driver_name, "net_i40e") == 0 && RTE_VER_YEAR <= 19)
```
I've moved it to util-dpdk-i40e.[ch]

lukashino commented on src/util-dpdk-i40e.c:
```diff
@@ -0,0 +1,348 @@
+/* Copyright (C) 2021 Open Information Security Foundation
```
done

lukashino commented:
> > For the record of i40e RSS configuration evaluation. Victor have noticed that I set IP but also SCTP / TCP/ UDP/ OTHER and fragmented IP patterns (e.g. function `i40eDeviceSetRSSFlowIPv4`). Since I haven't tested receiving fragmented packets when I first submitted this PR, I've decided to give it a try. Tests:
> > 
> > 1. As is in PR:
> > 
> > 
> > * fragmented and nonfragmented packets are received by the same workers even when I applied seed to alter them via `tcpreplay-edit` (option `--seed`)
> > 
> > 
> > 1. Setting only `ETH_RSS_FRAG_IPV4` and `ETH_RSS_IPV4`
> > 
> > 
> > * when setting `ETH_RSS_IPV4`, the PMD driver says that pctype is not supported (generally this means that the "type" of traffic is not a valid configuration for the i40e)
> > 
> > 
> > 1. Setting only `ETH_RSS_FRAG_IPV4` and `ETH_RSS_NONFRAG_IPV4_OTHER`
> > 
> > 
> > * this doesn't work well, packets of the same flow are received on the different workers (my explanation is that the fragmented packets are matched with `ETH_RSS_FRAG_IPV4` but the other UDP packets are not matched with `ETH_RSS_NONFRAG_IPV4_OTHER` rte_flow rule.
> > 
> > Also only IPV4 version of flows were mentioned but I applied the same settings to IPv6 rules.
> > For this reason, I believe the configuration is correct and the RSS distributes packets as it should.
> 
> Can you make sure this analysis is part of the git commit messages for the i40e specific commit?

Analysis included in the i40e RSS commit message

lukashino commented:
Just to provide some update, I have included some of @victorjulien commits in my branch (cherry picked), I have abstracted the configuration of RSS hash functions (for ixgbe (DPDK 19) and i40e (DPDK 19 and 20)). The hash functions for DPDK 19 were taken from your commit, I obtained hash functions for DPDK v20 for i40e.  So far, it's in my v9 branch.

lukashino commented:
Closing, let's continue in #6652
